### ISSUE TITLE
Fl_Input in an Fl_Pack in an Fl_Scroll unexpected behavior

### ISSUE THREAD

**Issue by MoAlyousef** [fltk/fltk#1172]
**Describe the bug**
Hello, a similar [issue](https://github.com/fltk-rs/fltk-rs/discussions/1616) was reported on the Rust bindings.
Placing multiple inputs inside a pack which is inside a scroll, then scrolling downwards and inserting text into one of the inputs, the change would only show after a redraw event such as scrolling back up and down again.

**To Reproduce**
```c++
#include <FL/Fl.H>
#include <FL/Fl_Double_Window.H>
#include <FL/Fl_Scroll.H>
#include <FL/Fl_Pack.H>
#include <FL/Fl_Input.H>

int main() {
    int x = 0, y = 0, w = 400, h = 300;
    auto *win = new Fl_Double_Window(w, h);
    auto *scroll = new Fl_Scroll(x, y, w, h, "Group 1");
    auto *pack = new Fl_Pack(x, y, w, h);
    pack->type(Fl_Pack::VERTICAL);
    for (auto i = 0; i < 20; i++) new Fl_Input(0, 0, 0, 30);
    pack->end();
    scroll->end();
    win->end();
    win->show();
    return Fl::run();
}
```

1. Run the app
2. Scroll downwards to one of the lower input fields, try to input text.
3. Observe that no text appears, neither does the cursor for that matter.
4. Scroll back up then down, the text appears.

**Expected behavior**
A user shouldn't have to scroll back up then down to see the text. It should appear immediately.

**Screenshots**
N/A

**FLTK Version**
Please complete the following information and delete non-applicable lines:
 - Version: 1.4.1
 - Downloaded and self-built from: Git
 - If from Git, branch: master

**FLTK Configure / Build Options**
`cmake -Bbin -GNinja -DCMAKE_BUILD_TYPE=Debug && cmake --build bin`

**Operating System / Platform:**
Please be as precise as possible, e.g. "Linux: Ubuntu 20.04"
 - OS: Ubuntu 22.04 in WSL.

**Linux/Unix Runtime, if applicable:**
 - X11

**Additional context**
N/A

**Comment by abau** [comment]
(Author of the original report in the `fltk-rs` repository)

I use KDE Plasma 6.2.4 under Arch Linux.

**Comment by erco77** [comment]
I can replicate here on Ubuntu 20.04 with fltk 1.4.x current.

It seems to be a redraw problem; typing doesn't show but it's there.. scrolling back and forth makes some of the text appear, but it still doesn't draw quite right unless I scroll the fields off the screen then back again.

I haven't checked the OP's code for problems, but I can definitely replicate. Here's a screenshot after scrolling down, clicking in a few fields and typing 123123 in each. At first nothing appears, but if I use the horiz scroll to scroll side to side, I get this partial redraw result.

![image](https://github.com/user-attachments/assets/c95cb0af-ac20-4864-91b4-fc511cad2047)

**Comment by erco77** [comment]
This seems to be a regression, as it's not a problem in 1.3.x.

**Comment by erco77** [comment]
One thing that seems to matter, not sure why: to replicate I had to partially scroll down (1/2 way, not all the way).

If I scroll to the _very bottom_ and click on the _last_ field I can type into it and it works fine.

But if I close and rerun the app, scroll to the very bottom and click on the _second to last_ field, I can't see what I'm typing. But if I then click on the last line, I can type normally. I can then click back onto the lines above that didn't work previously, and type new text and that appears normally. Strange.

**Comment by MatthiasWM** [comment]
It's happening on macOS as well. I'll have a stab at it.

**Comment by MatthiasWM** [comment]
Ok, I will try to explain what went wrong. It's basically a series of optimizations that eventually canceled each other out. There is no specific commit that breaks things, but it's a mix of deferring the Fl_Pack children layout, and optimizing Fl_Scroll.

In Fl_Input (or any other widget), when the content changes and the widget needs a redraw, it must notify the parents up the hierarchy until it finally creates a damaged area message to the underlying OS/Window Manager. To save time, if the FL_DAMAG_ALL flag is already set, the widget assumes that a redraw is pending, and will not ask for another redraw until the flag is cleared. Make sense.

If Fl_Pack changes position (in this case, is moved by Fl_Scroll), it will trigger a redraw event which in turn will eventually call Fl_Pack::draw() which in turn will rearrange all children. Now if a child was moved due to the relayout, Fl_Pack simply sets FL_DAMAGE_ALL on the child and the calls optimized draw functions on the child, relying on draw() to clear FL_DAMAGE_ALL. Seems to make sense (well, "relies on" is the issue here).

Now Fl_Scroll typically only shows a small part of all the children it has. So it calls fl_push_clip to render only the visible area.

In conclusion, if Fl_Pack sets the FL_DAMAGE_ALL flag of an invisible child widget that is then optimized out of the render loop because its graphics are clipped, FL_DAMAGE_ALL will not be cleared and redraw request will not propagate to the window and changes are not visible.

My patch makes sure that every time Fl_Pack will set FL_DAMAGE_ALL, it will also clear it, even if the widget is not drawn. This is not perfect and I can construct cases where FL_DAMAGE_ALL may have been set for another reason, but in this combination, revealing the widget later will trigger the required redraw. I will think about this a bit more, but so far, this patch should work.

**Comment by MoAlyousef** [comment]
The fix works on my end and I've updated fltk-rs accordingly. @abau can you try using:
```toml
[patch.crates-io]
fltk = { git = "https://github.com/fltk-rs/fltk-rs" }
```

**Comment by abau** [comment]
Works for me, too. Thanks everyone!
